components/net/image_cache.rs

```diff
@@ -559,11 +527,20 @@ impl ImageCacheStore {
 
     /// Insert received keys into the cache and complete the loading of images
     fn insert_keys_and_load_images(&mut self, image_keys: Vec<WebRenderImageKey>) {
-        for image_key in image_keys {
-            if let Some(pending_key) = self.key_cache.no_key_yet.pop_front() {
-                self.set_key_and_finish_load(pending_key, image_key);
-            } else {
-                self.key_cache.add_key(image_key);
+        if let KeyCacheState::Ready(ref mut queue) = self.key_cache.cache {
```

Comment:
I think you should turn this on it's head, by doing `if let KeyCacheState::PendingBatch`, and in the `else` case, do `unreachable!("A batch was received while we didn't request one")` because it's never supposed to happen.